docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,338 @@
+
+This document will propose the solution for above problem and provide the details of implementing it.
+
+## Solution Requirements
+
+A solution for this problem should :
+
+1. Update the desired state of the K8s custom resource with late initialized fields which are optional and not provided by the K8s user. 
+
+2. Provide configuration for AWS service teams to indicate which operation output returns the late initialized fields. i.e. Get/List/Create/Update output.
+
+    > AWS APIs have different behavior for returning the late initialized fields to the user. Some services return these fields as part of Create output, while some services will return the late initialized fields as part of  ReadOne output after the resource creation.
+
+3. Wait for asynchronous resource creation to populate late initialized fields.
+
+    > For some services/resources the late initialized fields do not appear immediately.  This happens when there is asynchronous processing involved in resource creation. The proposed solution should allow waiting and retry until the late initialized fields appear in the mentioned API output.
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Proposed Solution
+In this solution, There will be updates in 
+1) **ACK Code Generator** to capture the late initialized fields from AWS operation output and use them to late initialize k8s custom resource object. Code generator changes will support a) configuration for service teams to tell ACK code generator which fields will be late initialized and b) configuration to provide custom code for late initialization.
+2) **ACK Runtime** to a) update the custom resource stored in etcd with late initalized optional spec fields And b) handle the late initalized fields for resources that have asynchronous creation.
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. **Provide a new "ServerSideDefaultsConfig" struct for AWS service teams to provide which spec fields(spec.A) or spec field members(spec.A.X) will be late initialized.**
+
+    Only the fields/members mentioned as part of 'ServerSideDefaultsConfig' will be considered for late initialization. (The reason is explained inside step 3 "NOTE" )
+
+    `ServerSideDefaultsConfig` struct will be the member of `ResourceConfig` and it will look like following:
+
+    ```
+    type ServerSideDefaultsConfig struct { 
+    
+    // valid values: 'readOne, readMany, create, update'
```

Comment:
For Create and Update, we do NOT set spec fields in the latest object because [this](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/generate/code/set_resource.go#L93) parameter is true only for sdkFind operation. I have mentioned the reason in the proposal as well.

The reason is because Spec is created from CreateInput shape and there can be discrepancy of member shape if we try to construct/set spec fields from CreateOutput shape. 
One example is apigatewayv2 Integration resource's TLSConfig member. TLSConfig name is same in CreateInput and CreateOutput but their shapes are different.

Because of the above reason, I added Create and Update source in the above config. So that after reading this config, AWSResourceManager Create and Update will only set defaults for fields mentioned by service teams, not all spec fields.

Reply:
> For Create and Update, we do NOT set spec fields in the latest object because [this](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/generate/code/set_resource.go#L93) parameter is true only for sdkFind operation. I have mentioned the reason in the proposal as well.
> 
> The reason is because Spec is created from CreateInput shape and there can be discrepancy of member shape if we try to construct/set spec fields from CreateOutput shape.
> One example is apigatewayv2 Integration resource's TLSConfig member. TLSConfig name is same in CreateInput and CreateOutput but their shapes are different.

I think we need to revert that code and deal with the whole member field shape mismatch between Input and Output problem separately from whether we actively set the value of a Spec field to the value in a Create or Update Output shape.